### PR TITLE
[WIP] Adding the MAT Featurizer 

### PR DESCRIPTION
**Pull request by atreyamaj**
## Description
Currently working on the MAT Featurizer. There are some issues which I am looking into. Will add documentation after code review.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:17]: Lines 15-26 might not be needed because the `TorchModel` class already has an argument `device`, and handles the device change while preparing input batches for the model.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:40]: There's already a `one_hot_encode` function in `deepchem.utils.molecule_feature_utils`, use that one.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:65]: `adjacency_matrix = Chem.rdmolops.GetAdjacencyMatrix(mol)`. A function in rdkit already exists for adjacency matrix creation.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Done! I've removed the Cuda declaration lines.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Done! I've replaced all OHVector calls with the Deepchem one_hot_encode function and deleted the OHVector function.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Done! I have also replaced the distance matrix calculation with the inbuilt rdkit.Chem.molops.GetDistanceMatrix(mol) function. I did not know about these earlier, thank you!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:29]: `atom.GetAtomicNumber` I think

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:45]: `add_dummy_node` and `one_hot_formal_charge` are already class attributes. You can just use `self.add_dummy_node` instead?

**Comment by vsomnath** [comment]
Left some comments, I will do a more thorough pass later. Also add a test as you are writing code.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:29]: I just checked the Rdkit docs, it says getAtomicNum [here](https://www.rdkit.org/docs/source/rdkit.Chem.rdchem.html)

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Done!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: It’s GetAtomicNum instead of getAtomicNum

**Comment by atreyamaj** [comment]
> Left some comments, I will do a more thorough pass later. Also add a test as you are writing code.

Thank you, I have made the corresponding changes. I will work on the other aspects until then!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Good catch! Thank you, I have pushed the new change now.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:28]: DeepChem already has `atom_features` inside `deepchem/feat/graph_features.py`. Add a docstring saying that you use MAT specific features here. Also, write a docstring and test for everything as you go along, and not after you have written the full function. I understand things change, but it is good practice to write them parallely, and one needs to make a habit out of it.

Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:28]: Thank you, I will keep this in mind and also add the docstring/s.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:4]: We should move the numpy/pandas/sklearn declarations outside the try/catch block (only optional dependencies go in there)

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:19]: Can you expand this docstring with an explanation of what this featurizer does, an example of usage, and a link to the original paper?

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:23]: Can you add type signatures here?

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:30]: The parameter docstring isn't formatted correctly here

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:40]: Can you explain more about what features you are adding that are new? We will also need a parameter docstring added here

**Review by rbharath — COMMENTED** [review]
Just did a first pass! Mostly minor comments so far

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Done! I've removed the entire try-except block and gotten rid of unnecessary dependencies.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: I've added the others, however I'm not sure where I should add the usage example

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Done!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Fixed it, thanks!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:45]: Done! Please take a look.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:34]: Can you move the reference section to the main docstring for the class?

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:12]: To see how to add examples, see https://numpydoc.readthedocs.io/en/latest/format.html#examples. You'll want to do something like

```
Examples
--------
>>> feat = MATFeaturizer(...)
>>> out = feat.featurize("CCC")
```

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:20]: This should be type `RDKITMol`. We can't directly use rdkit types here since it's an optional dependency. Check out the type signatures of the other featurizers for examples

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:45]: Ok looks good I think

**Review by rbharath — COMMENTED** [review]
Did a second review pass

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:34]: Done!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:20]: Done!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:13]: Done! Please have a look.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:31]: `add_dummy_node` is not used anymore, so it should be removed. I guess this goes in the batch preparation code.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: Why do we resize `adjacency_matrix` and `distance_matrix` to `node_features.shape`?
I looked at the docs https://numpy.org/doc/stable/reference/generated/numpy.ndarray.resize.html, and I am confused by the example under "Enlarging an array: as above, but missing entries are filled with zeros". I would ideally expect zeros to be added to both rows, but this flattens the matrix, adds zeros, and then reshapes. This is not the desired behavior, I think.

File: deepchem/feat/tests/test_mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/tests/test_mat_featurizer.py:26]: shape and type tests are not the only ones you should be checking for. As @rbharath mentioned on the call, you should also compute the outputs for simple examples, and see if what the featurizer gives matches with what is the computed output.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:89]: `Tuple[np.ndarray]`

**Review by vsomnath — CHANGES_REQUESTED** [review]

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: Your adjacency and distance matrices have shape (N, N), and the node features matrix has shape (N, F). How would resize work if you have a large molecule where N > F?

Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: Yes I had some concerns about this. However, returning the arrays together or as a numpy object doesn't work. I am not sure how else to return them. :(

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:89]: Good catch! Thank you.

Comment (vsomnath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:89]: Ofcourse this depends on the final return statement. Can be adapted

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: `resize` is the incorrect method. Worst case, you should just concatenate the three matrices to get something of shape (N, 2N+F). Maybe the problem arises because `MATFeaturizer` is treated as a molecule featurizer when it is closer to a graph featurizer.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:31]: +1

File: deepchem/feat/tests/test_mat_featurizer.py
Comment (rbharath) [inline deepchem/feat/tests/test_mat_featurizer.py:8]: Some formatting issues here and in other docstrings

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (rbharath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: I'm a little confused by the resizing here. Do we expect to add these matrices together in the model? If not, then we can probably just leave them as different sizes right?

Sorry this might be a basic thing I'm missing here :)

**Review by rbharath — COMMENTED** [review]
Added on a couple of comments below.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: Me too. Worst case of returning a tuple is causing problems downstream, one can just do concatenation as mentioned above.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:31]: Removed!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: I think the authors split the dataset into 2 parts post featurization: X had the descriptors and Y had the labels. Thus, concatenation can work I believe. I'll go ahead with this for now then.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:36]: Remove `add_dummy_node` from the docstring too!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:94]: Unnecessary spacing I think

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: This is not needed. Node features has shape (N, F), Adjacency matrix has shape (N, N) and Distance Matrix has shape (N, N), where N is the number of atoms. 

```result = np.concatenate([node_features, adjacency_matrix, distance_matrix], axis=1)
```
is all you need to do.

File: deepchem/feat/tests/test_mat_featurizer.py
Comment (vsomnath) [reply deepchem/feat/tests/test_mat_featurizer.py:8]: The formatting issues still persist (the triple quotes need to be aligned). But I wonder if `yaml` is responsible for this. I have seen this happen when I used to format too.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:99]: Thank you, this worked! I tried using concatenate before but was unsuccessful for some reason.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:94]: Removed!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:36]: Done!

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: I think you might have missed passing the arrays as a list or tuple.

File: deepchem/feat/tests/test_mat_featurizer.py
Comment (atreyamaj) [reply deepchem/feat/tests/test_mat_featurizer.py]: Fixed! Thank you.

**Comment by vsomnath** [comment]
`RDKit` is a soft dependency. Follow this as an example for the import. https://github.com/deepchem/deepchem/blob/851ea515b7d847c81136ac6287b1456799b7f78f/deepchem/feat/molecule_featurizers/atomic_coordinates.py

This should help fix one of the failures.

**Comment by atreyamaj** [comment]
> `RDKit` is a soft dependency. Follow this as an example for the import. https://github.com/deepchem/deepchem/blob/851ea515b7d847c81136ac6287b1456799b7f78f/deepchem/feat/molecule_featurizers/atomic_coordinates.py
> 
> This should help fix one of the failures.

Thank you. I have made the change. I have also added function return type annotations.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py
Comment (vsomnath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:87]: np.ndarray now. Just mention concatenated version of nod features, adjacency and distance matrix.

Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Done! Please have a look.

**Comment by vsomnath** [comment]
Looks fine to me now. I think there are some flake8 and mypy issues that need to be fixed before this can be merged.

**Comment by atreyamaj** [comment]
> Looks fine to me now. I think there are some flake8 and mypy issues that need to be fixed before this can be merged.

Thank you for the heads-up! I have resolved the flake8 issues. I checked the mypy error log however there are none from the files that have been changed in the PR.

**Review by vsomnath — APPROVED** [review]

**Comment by vsomnath** [comment]
The test corresponding to MAT featurizer is failing. I hope you tested it locally before pushing.

**Comment by atreyamaj** [comment]
> The test corresponding to MAT featurizer is failing. I hope you tested it locally before pushing.

I did! As a precaution I'll check again and get back to you in a bit, I am not at home currently.

**Review by rbharath — APPROVED** [review]
Looks good on my end. Once the issues that @vsomnath has raised are addressed we should be good to merge in

**Comment by atreyamaj** [comment]
> Looks good on my end. Once the issues that @vsomnath has raised are addressed we should be good to merge in

Apologies, it looks like my push wasn't the latest one. I have updated it, there should be no more pytest errors. 

@vsomnath @rbharath

**Pull request merged**
